ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -415,19 +448,39 @@ static int st_object_record_update(st_data_t key, st_data_t value, st_data_t ext
 
   if (!ruby_ref_from_id(LONG2NUM(obj_id), &ref)) {
     // Id no longer associated with a valid ref. Need to delete this object record!
-
-    // Starting with the associated heap record. There will now be one less tracked object pointing to it
-    heap_record *heap_record = record->heap_record;
-    heap_record->num_tracked_objects--;
-
-    // One less object using this heap record, it may have become unused...
-    cleanup_heap_record_if_unused(recorder, heap_record);
-
-    object_record_free(record);
+    on_committed_object_record_cleanup(recorder, record);
     return ST_DELETE;
   }
 
-  // If we got this far, entry is still valid
+  // If we got this far, then we found a valid live object for the tracked id.
+
+  #ifdef HAVE_WORKING_RB_GC_FORCE_RECYCLE
+    // If we are in a ruby version that has a working rb_gc_force_recycle implementation,
+    // its usage may lead to an object being re-used outside of the typical GC cycle.
+    //
+    // This re-use is in theory invisible to us and would mean that the ref from which we
+    // collected the object_record metadata may not be the same as the current ref and
+    // thus any further reporting would be innacurately attributed to stale metadata.
+    //
+    // In practice, there is a way for us to notice that this happened because of a bug
+    // in the implementation of rb_gc_force_recycle. Our heap profiler relies on object
+    // ids and id2ref to detect whether objects are still alive. Turns out that when an
+    // object with an id is re-used via rb_gc_force_recycle, it will "inherit" the ID
+    // of the old object but it will NOT have the FL_SEEN_OBJ_ID as per the experiment
+    // in https://github.com/DataDog/dd-trace-rb/pull/3360#discussion_r1442823517
+    //
+    // Thus, if we detect that the ref we just resolved above is missing this flag, we can
+    // safely say re-use happened and thus treat it as an implicit free of the object
+    // we were tracking (the original one which got recycled).
+    if (RB_FL_ABLE(ref) && !RB_FL_TEST(ref, RUBY_FL_SEEN_OBJ_ID)) {
+      on_committed_object_record_cleanup(recorder, record);
+      return ST_DELETE;
+    }
+
+    // NOTE: The bug mentioned above may also lead to a slow leaking of records in the
+    // objspace obj_to_id and id_to_obj tables. We could try also force-setting the seen
+    // obj id bit here to reduce the impact of this leak.
```

Comment:
I left this note here because there are 2 places where we could set the bit:
* When we sample a new allocation on a recycled slot. Here we have to set a bit otherwise our tracking breaks.
* When we detect that an object we were tracking got recycled into a new object for which we didn't get an allocation. This is the situation where this note is. We don't really need to set the bit here for our heap profiler to work. We could set it to further improve leak impact but it could be argued it would be an extra unnecessary meddling with runtime internals.

Since we're already setting in that other place, I'd vote to also set it in this codepath but wanted to run that by you first and see if you shared this opinion 😄

Reply:
Did so and added a test for it.